### PR TITLE
fix(integrations): make LightGBM callback compatible with 4.0.0

### PR DESCRIPTION
**Pull request by ayulockin**
Fixes
-----
- Fixes [WB-14525](https://wandb.atlassian.net/browse/WB-14525)

Description
-----------
This PR makes our LightGBM callback compatible with the latest release of lightgbm 4.0.0. The changes are backwards compatible as well - tested with 3.3.5.

### <samp>🤖 Generated by Copilot at 0238ecc</samp>

Refactor LightGBM integration to use a new callback class for logging to W&B. The new `_WandbCallback` class handles the logic of tracking metrics and parameters, and the `wandb_callback` function becomes a thin wrapper that creates and returns the callback object. This improves the code readability and consistency with other integrations.

Testing
-------
How was this PR tested?

Checklist
-------
- [x] Include reference to internal ticket "Fixes WB-NNNNN" and/or GitHub issue "Fixes #NNNN" (if applicable)
- [x] Ensure PR title compliance with the [conventional commits standards](https://github.com/wandb/wandb/blob/main/CONTRIBUTING.md#conventional-commits)

### <samp>🤖 Generated by Copilot at 0238ecc</samp>

> _`_WandbCallback` class_
> _logs LightGBM results and params_
> _simpler, consistent_

[WB-14525]: https://wandb.atlassian.net/browse/WB-14525?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by dmitryduev — APPROVED** [review]

**Comment by kptkin** [comment]
@ayulockin Thanks for the fix!

one small follow up, could you please also unpin [this test](https://github.com/wandb/wandb/blob/main/tests/functional_tests/t0_main/lightgbm/t1_regression.yea#L9-L10) (to make sure your fix is tested)? and also add another version of a test for a version pre 4.0?

**Review by kptkin — CHANGES_REQUESTED** [review]
see my comment above :)

**Comment by ayulockin** [comment]
@kptkin can you confirm if the changes are good.

File: tests/functional_tests/t0_main/lightgbm/t1_regression.py [hunk 1]
```diff
@@ -7,6 +7,8 @@
 import wandb
 from wandb.integration.lightgbm import log_summary, wandb_callback
 
+print(lgb.__version__)
```
Comment (kptkin) [inline tests/functional_tests/t0_main/lightgbm/t1_regression.py:10]: remove print when you are done debugging

File: tests/functional_tests/t0_main/lightgbm/t2_regression.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
+#!/usr/bin/env python
+"""Test lightgbm integration."""
+
+import lightgbm as lgb
+import pandas as pd
+import requests
+import wandb
+from wandb.integration.lightgbm import log_summary, wandb_callback
+
+print(lgb.__version__)
```
Comment (kptkin) [inline tests/functional_tests/t0_main/lightgbm/t2_regression.py:10]: oh you want to keep it as part of the test... i don't mind either way

**Review by kptkin — APPROVED** [review]
Thanks for adding the extra test

File: tests/functional_tests/t0_main/lightgbm/t2_regression.py (hunk 2 above)
Comment (ayulockin) [reply tests/functional_tests/t0_main/lightgbm/t2_regression.py:10]: Yeah I wanted to keep it as part of the test.

**Pull request merged**